Add pointer concept (#1724)
exercism/go: concepts/pointers/about.md

var p *int
p = &a // the variable 'p' will hold the memory address of 'a'
```

### Accessing the value via a pointer (dereferencing)

When we have a pointer, we might want to know the value stored in the memory address to which it points. We can do this using the `*` operator:

```go
var a int 
a = 2

var p *int 
p = &a // the variable 'p' will hold the memory address of 'a'

var b int
b = *p // b == 2
```

The operation `*p` fetches the value stored at the memory address stored in `p`. This operation is often called "dereferencing".

We can also use the derefering operator to assign a new value to the memory address referenced by the pointer:

```go
var a int
a = 2   // declare int variable 'a' and assign it the value of 2

var pa *int
pa = &a          // 'pa' now refers to the memory address of 'a'  
*pa = *pa + 2    // increment by 2 the value at memory address 'pa'

fmt.Println(a)   // Output: 4
                 // 'a' will have the new value that was changed via the pointer!
```

Assigning to `*pa` will change the value stored at the memory address `pa` holds. Since `pa` holds the memory address of `a`, by assigning to `*pa` we are effectively changing the value of `a`! 

> A note of caution however: always check if a pointer is not `nil` before dereferencing. Dereferencing a `nil` pointer will make the program crash at runtime!

Reviewer: I think it is very important to show the actual error message here so people can recognize it later when it happens to them.
```suggestion
A note of caution however: always check if a pointer is not `nil` before dereferencing. Dereferencing a `nil` pointer will make the program crash at runtime!

```go
var p *int // p is nil initially
fmt.Println(*p)
// panic: runtime error: invalid memory address or nil pointer dereference
```

```go
var p *Person
p = &Person{Name: "Peter", Age: 22}

fmt.Println(p.Name) // Output: "Peter" 
                    // Go automatically dereferences 'person' to allow
                    // access to the 'Name' field
```

## Slices and maps are already pointers

Slices and maps are special types because they already have pointers in their implementation. This means that more often that not, we don't need to create pointers for these types to share the memory address for their values. Imagine we have a function that adds a key to a map:

```go
func addPeterAge(m map[string]int){
	m["Peter"] = 22
}
```

If we create a map and call this function, the changes the function made to the map persist after the function ended. This is a similar behavior we get if we were using a pointer, but note how on this example we are not using any referencing/dereferencing or any of the pointer syntax:

```go
ages := make(map[string]int)
addPeterAge(ages)
fmt.Println(ages)
// Output: map[Peter:22]
// The changes the function 'addPeterAge' made to the map are visible after the function ends!
```

> 

Reviewer: Suggestion how to address the slices issue mentioned here https://github.com/exercism/go/pull/1724#issuecomment-932351399
```suggestion

The same applies when changing an existing item in a slice. However, actions that return a new slice like `append` will **not** modify the slice outside of the function.
```

Reviewer 2: This is actually a common edge case in go.
```
func addOne(in []int) []int {
	return append(in, 1)
}

func main() {
	slice := []int{1, 2, 3, 4}
	slice = append(slice, 5)
	slice2 := append(slice, 6)
	fmt.Println(addOne(slice)) // prints 1,2,3,4,5,1
	fmt.Println(slice2)        // also prints 1,2,3,4,5,1
}
```

## Pointer arithmetic

> Unlike other languages, pointer arithmetic is not allowed in Go. This means that snippets like the following are invalid:

Reviewer: Suggesting to add an explanation of that pointer arithmetic actually is. Also rephrased the first bit a little.
```suggestion
Unlike other languages, Go does not support pointer arithmetic. This means you cannot go from a memory address of one value to the address of another by incrementing the pointer value. So snippets like the following are invalid:
```

```go
a := 2
p := &a // the variable 'p' wil hold the memory address of 'a'

b := *p // b == 2
```

The operation `*p` fetches the value stored at the memory address stored in `p`. This operation is often called "dereferencing".

We can also use the derefering operator to assign a new value to the memory address referenced by the pointer:

```go

a  := 2          // declare int variable 'a' and assign it the value of 2 
pa := &a         // 'pa' now refers to the memory address of 'a'  
*pa = *pa + 2    // increment by 2 the value at memory address 'pa'

fmt.Println(a)   // Output: 4
                 // 'a' will have the new value that was changed via the pointer!
```

Assigning to `*pa` will change the value stored at the memory address `pa` holds. Since `pa` holds the memory address of `a`, by assigning to `*pa` we are effectively changing the value of `a`! 

A note of caution however: always check if a pointer is not `nil` before dereferencing. Dereferencing a `nil` pointer will make the program panic at runtime!

### Pointers to structs

So far we've only seen pointers to primitive values. We can also create pointers for structs:

```go

> type Point struct {

Reviewer 3: I'd personally recommend using another name for this struct type as the `Point` name is too similar to the pointer concept and may confuse readers. Perhaps a `Person` struct instead?

Author: That's a good point. After writing it down I also thought about that. I'll change it :)

Reviewer 3: Sorry to butt in! I was just poking around at the open pull requests. Appreciate your work here!

# About

Like many other languages, Go has pointers. 
If you're new to pointers, they can feel a little mysterious but once you get used to them, they're quite straight-forward.
They're a crucial part of Go, so take some time to really understand them.

Before digging into the details, it's worth understanding the use of pointers. Pointers are a way to share memory with other parts of our program, which is useful for two major reasons:
1. When we have large amounts of data, making copies to pass between functions is very inefficient. 
  By passing the memory location of where the data is stored instead, we can dramatically reduce the resource-footprint of our programs.

> 2. By passing pointers between functions, we can access and modify the computer's memory directly, meaning that any changes made by one function are immediately visible to other parts of the program when the function ends.

Reviewer 2: s/computer's memory/single copy of the data/

A note of caution however: always check if a pointer is not `nil` before dereferencing. Dereferencing a `nil` pointer will make the program crash at runtime!

```go
var p *int // p is nil initially
fmt.Println(*p)
// panic: runtime error: invalid memory address or nil pointer dereference

### Pointers to structs

So far we've only seen pointers to primitive values. We can also create pointers for structs:

```go
type Person struct {
    Name string
    Age  int
}

var peter Person
peter = Person{Name: "Peter", Age: 22}

var p *Person
p = &peter
```

We could have also created a new `Person` and immediately stored a pointer to it:

```go
var p *Person
p = &Person{Name: "Peter", Age: 22}
```

When we have a pointer to a struct, we don't need to dereference the pointer before accessing one of the fields:

```go
var p *Person
p = &Person{Name: "Peter", Age: 22}

fmt.Println(p.Name) // Output: "Peter" 
                    // Go automatically dereferences 'person' to allow
                    // access to the 'Name' field
```

## Slices and maps are already pointers

> Slices and maps are special types because they already have pointers in their implementation. This means that more often that not, we don't need to create pointers for these types to share the memory address for their values. Imagine we have a function that adds a key to a map:

Reviewer 2: This is actually a potential edge case:
```
func addOne(in []int) []int {
	return append(in, 1)
}

func main() {
	slice := []int{1, 2, 3, 4}
	slice = append(slice, 5)
	slice2 := append(slice, 6)
	fmt.Println(addOne(slice)) // prints 1,2,3,4,5,1
	fmt.Println(slice2)        // also prints 1,2,3,4,5,1
}
```

```go
func addPeterAge(m map[string]int){
	m["Peter"] = 22
}
```

If we create a map and call this function, the changes the function made to the map persist after the function ended. This is a similar behavior we get if we were using a pointer, but note how on this example we are not using any referencing/dereferencing or any of the pointer syntax:

```go
ages := make(map[string]int)
addPeterAge(ages)
fmt.Println(ages)
// Output: map[Peter:22]
// The changes the function 'addPeterAge' made to the map are visible after the function ends!
```

> The same applies when changing an existing item in a slice. However, actions that return a new slice like `append` will **not** modify the slice outside of the function.

Reviewer 2: This is not necessarily true:
```
func addOne(in []int) []int {
	return append(in, 1)
}

func main() {
	slice := []int{1, 2, 3, 4}
	slice = append(slice, 5)
	slice2 := append(slice, 6)
	fmt.Println(addOne(slice)) // prints 1,2,3,4,5,1
	fmt.Println(slice2)        // also prints 1,2,3,4,5,1
}
```

Author: Maybe we can say "might not modify the slice" and link to a resource that explains this in more detail like [Arrays, slices (and strings): The mechanics of 'append'](https://go.dev/blog/slices).
